src/main/java/ProjectIndexer.java

```diff
@@ -58,7 +58,15 @@ public void index() {
         }
 
         for (DocumentIndexer indexer : indexers.values()) {
-            indexer.index();
+            indexer.preIndex();
+        }
+
+        for (DocumentIndexer indexer : indexers.values()) {
+            indexer.visitDefinitions();
+        }
+
+        for (DocumentIndexer indexer : indexers.values()) {
+            indexer.visitReferences();
```

Comment:
I split up `index()` into 3 pieces:

- `preIndex()` emits some metadata about the doc (similar in function to `postIndex()`)
- `visitDefinitions()`
- `visitReferences()`